Components/Queues/AsyncQueueConsumer.php

```diff
@@ -137,26 +128,36 @@ protected function process(EndpointInterface $queueEndpoint, MessageInterface $m
      * Overrides the main callback function to convert the AMQPMessage from the queue into a QueueMessage.
      *
      * {@inheritdoc}
-     *
-     * @return callable
      */
     protected function callback(EndpointInterface $endpoint): callable
     {
-        return function (AMQPMessage $message) use ($endpoint) {
+        return function (AMQPMessage $encodedMessage) use ($endpoint) {
             try {
                 $start = microtime(true);
-                $queueMessage = $this->serializer->deserialize($message->getBody(), SerializableInterface::class, $this->getQueueDriver($endpoint)->getFormat());
+                $message = $this->getSerializer()->decode([
+                    'body' => $encodedMessage->getBody(),
+                    'headers' => [],
+                ]);
+
                 $this->consumptionDuration = (microtime(true) - $start) * 1000;
 
-                $queueMessage->setMessageId($message->getDeliveryTag());
+                $message->setMessageId($encodedMessage->getDeliveryTag());
             } catch (\Exception $exception) {
+                $this->getExceptionHandler()->handle($exception, [
+                    'endpoint' => $endpoint,
+                    'body' => $encodedMessage->getBody(),
+                    'headers' => $encodedMessage->get('application_headers')->getNativeData(),
+                ]);
+                $message = new QueueMessage();
+                $message->setMessageId($encodedMessage->getDeliveryTag());
+                $this->getQueueDriver($endpoint)->ack($message);
+
                 $this->consumptionDuration = (microtime(true) - $start) * 1000;
-                $this->getExceptionHandler()($exception, ['message' => $message]);
 
-                return false;
+                return null;
```

Comment:
Would it make sense to expect something from the exception handler, and if returns `null`, consider the message poisoned and `ack` it? Otherwise it doesnt make sense to have this functionality when the built-in handler just rethorws the exception (and in essence, prevents this code from being executed).

Something like this (pseudocode)
```
if($message = handler->handle() === null) {
    ack()
    return null
}

parent::callback()
```

Reply:
Or even better (maybe it's an overkill), but trigger the handler wrapped in a `try/catch` that, if it catches a `PosionedMessageException`, acks it, otherwise continues with the execution.

Reply:
Changed to the option one